src/ray/common/cgroup/cgroup_utils.cc

```diff
@@ -0,0 +1,163 @@
+    OpenCgroupFileAndAppend(procs_path, absl::StrFormat("%d", ctx.pid));
+
+    const std::string max_memory_path = absl::StrFormat("%s/memory.max", cgroup_folder);
+    OpenCgroupFileAndAppend(max_memory_path, absl::StrFormat("%d", ctx.max_memory));
+  }
+
+  return true;
+}
+
+bool UpdateDefaultCgroup(const PhysicalModeExecutionContext &ctx) {
+  // Sanity check.
+  RAY_CHECK(!ctx.uuid.empty());
+  RAY_CHECK_EQ(ctx.uuid, kDefaultCgroupUuid);
+  RAY_CHECK_EQ(ctx.max_memory, 0);
+
+  const std::string cgroup_folder =
+      absl::StrFormat("%s/%s", ctx.cgroup_directory, ctx.uuid);
+  int ret_code = mkdir(cgroup_folder.data(), kCgroupFilePerm);
+  if (ret_code != 0) {
+    return false;
+  }
+
+  const std::string procs_path = absl::StrFormat("%s/cgroup.procs", cgroup_folder);
+  OpenCgroupFileAndAppend(procs_path, absl::StrFormat("%d", ctx.pid));
+
+  return true;
+}
+
+bool DeleteCgroup(const PhysicalModeExecutionContext &ctx) {
+  // Sanity check.
+  RAY_CHECK(!ctx.uuid.empty());
+  RAY_CHECK_NE(ctx.uuid, kDefaultCgroupUuid);
+  RAY_CHECK_GT(ctx.max_memory, 0);
+
+  const std::string cgroup_folder =
+      absl::StrFormat("%s/%s", ctx.cgroup_directory, ctx.uuid);
+  return rmdir(cgroup_folder.data()) == 0;
+}
+
+bool RemoveCtxFromDefaultCgroup(const PhysicalModeExecutionContext &ctx) {
+  // Sanity check.
+  RAY_CHECK(!ctx.uuid.empty());
+  RAY_CHECK_EQ(ctx.uuid, kDefaultCgroupUuid);
+  RAY_CHECK_EQ(ctx.max_memory, 0);
+
+  const std::string cgroup_folder =
+      absl::StrFormat("%s/%s", ctx.cgroup_directory, ctx.uuid);
+  int ret_code = mkdir(cgroup_folder.data(), kCgroupFilePerm);
+  if (ret_code != 0) {
+    return false;
+  }
+
+  const std::string procs_path = absl::StrFormat("%s/cgroup.procs", cgroup_folder);
+  std::ostringstream buffer;
+  {
+    std::ifstream file(procs_path.data(), std::ios::in);
+    buffer << file.rdbuf();
+  }
+  std::string content = buffer.str();  // contains all PIDs, separated by space
+
+  std::vector<std::string_view> old_pid_strings = absl::StrSplit(content, ' ');
+  std::vector<std::string_view> new_pid_strings;
+  new_pid_strings.reserve(old_pid_strings.size() - 1);
+  for (const auto &cur_pid : old_pid_strings) {
+    if (cur_pid == absl::StrFormat("%d", ctx.pid)) {
+      continue;
+    }
+    new_pid_strings.emplace_back(cur_pid);
+  }
+
+  const std::string new_pids = absl::StrJoin(new_pid_strings, " ");
+  {
+    std::ofstream out_file{procs_path.data(), std::ios::out};
+    out_file << new_pids;
+  }
+
+  return true;
+}
+
+}  // namespace
+
+bool SetupCgroupForContext(const PhysicalModeExecutionContext &ctx) {
+#ifndef __linux__
+  return false;
+#endif
+
+  // Create a new cgroup if max memory specified.
+  if (ctx.max_memory > 0) {
+    return CreateNewCgroup(ctx);
+  }
+
+  // Update default cgroup if no max resource specified.
+  return UpdateDefaultCgroup(ctx);
+}
+
+bool CleanupCgroupForContext(const PhysicalModeExecutionContext &ctx) {
+#ifndef __linux__
+  return false;
+#endif
```

Comment:
if you build for non-linux OS, does this constitute dead code after return? maybe you want an `#else` to capture the real impl.

Reply:
> if you build for non-linux OS, does this constitute dead code after return?

Yes, you're right; I didn't do that since I personally think wrapping code blocks inside of ifdef macro makes code a little un-readable.